use-cases/client-server-msgs-perf/svmsg_file_server.c

```diff
@@ -489,17 +539,38 @@ svr_proc_process_msg(void *cfg)
 
           case REQ_MT_EXIT:
 
-            clientp = &ActiveClients[req.client_idx];
-            // One client has informed that it has exited
-            NumActiveClients--;
+            // -ve client-index => it's a server-thread informing us to exit.
+            if (req.client_idx >= 0) {
+                clientp = &ActiveClients[req.client_idx];
+                // One client has informed that it has exited
+                NumActiveClients--;
+
+#ifdef __cplusplus
+                int activeClients = NumActiveClients.load();
+#else
+                int activeClients = NumActiveClients;
+#endif
+                printf("Server: Client ID=%d exited. num_ops=%" PRIu64 " (%s)"
+                       " # active clients=%d\n",
+                       req.clientId,
+                       clientp->num_ops, value_str(clientp->num_ops),
+                       activeClients);
+            }
 
-            printf("Server: Client ID=%d exited. num_ops=%" PRIu64 " (%s)"
-                   " # active clients=%d\n",
-                   req.clientId,
-                   clientp->num_ops, value_str(clientp->num_ops),
-                   NumActiveClients);
+            if (NumActiveClients <= 0) {
+                printf("Server: ThreadID=%" PRIu64 " exiting.\n", tid);
 
-            if (NumActiveClients == 0) {
+                if (NumActiveClients == 0) {
+                    req.mtype = REQ_MT_EXIT;
+
+                    // Inform last remaining server-thread that this is
+                    // a final-exit from a server-thread, and not from a real
+                    // client.
+                    req.client_idx = -1;
+                    if (msgsnd(serverId, &req, REQ_MSG_SIZE, 0) == -1) {
```

Comment:
This is a slightly hacky way of handling the case when last-client drains out of the system.

When that client sends the `REQ_MT_EXIT` message, the thread handling it will detect that `NumActiveClients` has gone down to 0.

It's likely that one of the other server threads is waiting for a message.

The code here "informs" the other thread to also exit. The receiving logic is above on L143.

---
This scheme works when tested with 2 or 4 threads and 5 clients.

It's not a general purpose fix, yet, but something to get tests working for now.
Once I get access to an AWS instance with more cores, will be able to re-test this exit condition handling with larger number(s) of clients & threads. And if necessary, will tighten this area to be robust.

Reply:
Would it be simpler just to `exit()` in this case? i.e. don't bother trying to wake the other threads, just nuke the whole the thing from orbit.

Reply:
We **do need** to invoke the `msgsnd()` to free-up the last remaining thread which is waiting for all clients to exit.

When that call to `msgsnd()` on line above fails, we do do a hard-exit(), so your suggestion is being done already.

Reply:
Ah yes, I see now. Understood.